### DISCUSSION TITLE
How to install on macOS with M1

### DISCUSSION THREAD

**Discussion by system1system2** [coqui-ai/TTS#2177] [General]
I struggled to install TTS on my MBP with M1 Pro as it's not officially supported. After much searching, thanks to previously published guidance from @jeffrafter, I managed to succeed. This is the procedure (assuming you have `miniconda` installed):

1. Go to the folder where you want to clone the TTS repo
2. `conda create --name coqui python=3.9` < you can replace "coqui" with whichever name you like
3. `conda activate coqui` < before that "coqui" here is the same name you decided for at step 2
4. `git clone https://github.com/coqui-ai/TTS.git`
5. `brew install mecab` < without this step, the "pip install" and "make install" commands will fail
6. `brew install espeak` < this is only necessary if you want to use the `tts_models/en/vctk/vits` voices
7. `conda install numpy scipy scikit-learn Cython`
8. `pip install -e .` < possibly unnecessary step as "make install" command should do the same. I didn't test without it.
9. `make install`

At this point, launch TTS or TTS-Server as indicated in the documentation.

**Comment by melMass** [comment]
Is there a way to avoid mecab all together? All my troubles with this project always are because of this library. I managed to make it work in the past, but after updating to Ventura it's broken again

**Reply by jeffrafter** [reply]
Look at https://github.com/coqui-ai/TTS/issues/2052. In general `mecab` works on Ventura. But this comment https://github.com/coqui-ai/TTS/issues/2052#issuecomment-1348481112 suggests this library could be reconfigured to not rely on it. I tried that and it worked but all of the standard out of the box models assumed `mecab` at the time I tried it. So I could work without it but it meant I had to retrain everything which I was avoiding at the time.

**Reply by melMass** [reply]
Thanks, custom training is the goal so I will check out to omit it thanks for the precision!

**Comment by kitschpatrol** [comment]
Thank you! This worked for me on M1 / Ventura.

Just one note for others: Without specifying a python version when creating the conda environment, mine ended up with 3.11 which did not work. Manually specifying 3.9 at environment creation did the trick:

`conda create --name coqui python=3.9`

**Reply by system1system2** [reply]
Very good suggestion. I updated my original post to incorporate your recommendation. Thank you

**Comment by cellulosa** [comment]
For anyone else having this issue:

    flake8 6.0.0 requires mccabe<0.8.0,>=0.7.0, but you have mccabe 0.6.1 which is incompatible.

I had to `brew uninstall mecab` and install that library with `python3 -m pip install mccabe`.

**Comment by munshaw** [comment]
I was trying to get Japanese TTS working on my Macbook Air M1 Sonoma, and ran into the same issues with mecab. The above solution wasn't working, but I found an alternative solution. It's not as fast, but I don't really need speed for simple TTS.

I created an emulated UTM VM for Debian 12 amd64, with 2GB of RAM. During the installer I didn't install any desktop, but did install the SSH server. After installing, I set the cpu to qemu64 + sse4.1 + sse4.2 + ssse3, deleted the display, and set up network to use Emulated VLAN and port forwarded TCP 22 to 22. I set up a paswordless ssh key to access the VM. Proceeded with installing TTS for Debian.

Then I created some shell scripts with ssh+scp in my path to generating TTS. I also ran fstrim, and compressed the disk image to save some space.

**Comment by tokamac** [comment]
> 3. conda activate coqui < before that "coqui" here is the same name you decided for at step 2
```
conda activate coqui
CondaError: Run 'conda init' before 'conda activate'
```
Off to a good start…

```
conda init
no change     /opt/homebrew/Caskroom/miniconda/base/condabin/conda
no change     /opt/homebrew/Caskroom/miniconda/base/bin/conda
no change     /opt/homebrew/Caskroom/miniconda/base/bin/conda-env
no change     /opt/homebrew/Caskroom/miniconda/base/bin/activate
no change     /opt/homebrew/Caskroom/miniconda/base/bin/deactivate
no change     /opt/homebrew/Caskroom/miniconda/base/etc/profile.d/conda.sh
no change     /opt/homebrew/Caskroom/miniconda/base/etc/fish/conf.d/conda.fish
no change     /opt/homebrew/Caskroom/miniconda/base/shell/condabin/Conda.psm1
no change     /opt/homebrew/Caskroom/miniconda/base/shell/condabin/conda-hook.ps1
no change     /opt/homebrew/Caskroom/miniconda/base/lib/python3.11/site-packages/xontrib/conda.xsh
no change     /opt/homebrew/Caskroom/miniconda/base/etc/profile.d/conda.csh
no change     /Users/user/.bash_profile
No action taken.
```
Then
```
conda activate coqui
CondaError: Run 'conda init' before 'conda activate'
```
What's wrong again right from the start!

**Reply by abdul-0** [reply]
@tokamac you must have a different shell, try `conda init zsh` for e.g

**Comment by Daniel3356** [comment]
Hey !
Is it fast on your machine ? I'm using a macbook pro M1 and I get 46 words per minutes.

**Comment by lukshan13** [comment]
Sorry for the necro. Despite following the steps, I'm still getting the same issue, "ERROR: Could not build wheels for TTS, which is required to install pyproject.toml-based projects"

I'm running M3 Pro on Sonoma 14.3. Would appreciate any help

**Reply by kitschpatrol** [reply]
Hey, I know "works on my machine" is never especially helpful, but I just went through the exact steps outline in the [OP](https://github.com/coqui-ai/TTS/discussions/2177?notification_referrer_id=NT_kwDOAAL2dLE0OTQzMTI1NzM0OjE5NDE2NA#discussion-4606735) [with the latest commit](https://github.com/coqui-ai/TTS/commit/dbf1a08a0d4e47fdad6172e433eeb34bc6b13b4e) of TTS, and it's working fine for me on an M3 Max with Sonoma 14.4... so it's probably down to a difference in your machine's configuration?

Have you run `brew doctor`? Are the Xcode command line tools installed?

**Reply by michaelmj** [reply]
I got this working today with M2 Max on 14.4 without needing any outside dependencies with the following steps:

1. Clone TTS repo.
2. `conda create --name TTS python=3.11`
3. `conda activate TTS`
4. `pip install -e .`

At that point `tts` works at the command line.

**Reply by rpoelking** [reply]
I've followed the steps with both the first post and this one from michaelmj. At the point of running ```pip install -e .```  I get an error saying that ```ERROR: file://Users/myname/current/path/ does not appear to be a Python project: neigher 'setup.py' nor 'pyproject.toml' found.``` Am I missing a dependency or am I supposed to change to a different directory. Where do I find setup.py?

Nevermind. I followed this link (https://blog.graywind.org/posts/coqui-tts-mac/) and the make now works.

**Comment by 3manu31** [comment]
I followed the steps but at the step: `brew install mecab` , I get: `zsh: command not found: brew`
(I am new to coding so I don't really get what's wrong)
Any help?

**Reply by rpoelking** [reply]
you need to install homebrew first. go to https://brew.sh/ for installation instructions. It's just a package manager for mac.

**Reply by 3manu31** [reply]
I thought i had it already installed, but i didn't. Thanks. Now it gives me this `Channels: - defaults
Platform: osx-arm64
Collecting package metadata (repodata.json): done
Solving environment: failed` 
in response to `conda install numpy scipy scikit-learn Cython` . 
Does this mean I need also Cython on my Mac?

**Comment by serendipity1137** [comment]
Let me know if this helps.

`/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`


```
# Install Conda
mkdir ~/coqui && cd ~/coqui # Or wherever you want to install Conda/Coqui
curl https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-arm64.sh -o miniconda3.sh
chmod +x ./miniconda3.sh
./miniconda3.sh
```

```
conda create --name coqui python=3.10 # Then hit y and enter to accept
conda activate coqui
git clone https://github.com/coqui-ai/TTS.git
# Install requirements
brew install mecab espeak
pip install numpy==1.21.6
conda install scipy scikit-learn Cython
# Install Coqui-TTS
cd TTS
make install
```

**Comment by Mudasirhabib123** [comment]
Following solution works on my M2 

$ python
Python 3.9.6 (default, Nov 10 2023, 13:38:27) 
[Clang 15.0.0 (clang-1500.1.0.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.get_include()
'/Users/mudasirhabib/Documents/projects/verticalsols/tts/venv/lib/python3.9/site-packages/numpy/core/include'
>>> quit()
export CFLAGS="-I /Users/mudasirhabib/Documents/projects/verticalsols/tts/venv/lib/python3.9/site-packages/numpy/core/include $CFLAGS"


And then 
pip install git+https://github.com/coqui-ai/TTS

**Comment by basudee** [comment]
I have managed to install this. Can someone please help me launch the WebUI? Thank you in advance

**Comment by sapardo10** [comment]
on M3 Max I was still getting this error:
`Compiling spacy/vectors.pyx because it changed.
      [ 1/41] Cythonizing spacy/attrs.pyx
      [ 2/41] Cythonizing spacy/kb.pyx
      Traceback (most recent call last):
        File "/opt/homebrew/anaconda3/envs/coqui/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/opt/homebrew/anaconda3/envs/coqui/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/opt/homebrew/anaconda3/envs/coqui/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/private/var/folders/ck/2dkp9skj33g3bm3dtppbl_980000gn/T/pip-build-env-1o5vpwar/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 332, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
        File "/private/var/folders/ck/2dkp9skj33g3bm3dtppbl_980000gn/T/pip-build-env-1o5vpwar/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 302, in _get_build_requires
          self.run_setup()
        File "/private/var/folders/ck/2dkp9skj33g3bm3dtppbl_980000gn/T/pip-build-env-1o5vpwar/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 318, in run_setup
          exec(code, locals())
        File "<string>", line 224, in <module>
        File "<string>", line 211, in setup_package
        File "/private/var/folders/ck/2dkp9skj33g3bm3dtppbl_980000gn/T/pip-build-env-1o5vpwar/overlay/lib/python3.9/site-packages/Cython/Build/Dependencies.py", line 1154, in cythonize
          cythonize_one(*args)
        File "/private/var/folders/ck/2dkp9skj33g3bm3dtppbl_980000gn/T/pip-build-env-1o5vpwar/overlay/lib/python3.9/site-packages/Cython/Build/Dependencies.py", line 1321, in cythonize_one
          raise CompileError(None, pyx_file)
      Cython.Compiler.Errors.CompileError: spacy/kb.pyx`

Followed this, and then it worked, not sure what exactly did the trick:

1. Ensure All Dependencies Are Installed
Since you're on macOS with M3 Max, some dependencies may need to be manually installed before building certain Python packages. Run the following commands:

```
brew install cython
brew install gcc
brew install openblas
```

You may also need Xcode command line tools. If you haven't installed them yet, run:

`xcode-select --install`

2. Use Pre-Compiled Binary Packages
If possible, install pre-compiled binary versions of packages, like spacy, instead of building them from source. You can try installing it with pip without building it:

`pip install --no-build-isolation --no-use-pep517 spacy`

3. Set Environment Variables
For compiling Cython extensions (used in packages like spacy), setting environment variables to help the compiler find the right headers and libraries may be necessary. Add these to your terminal before running the install command:

```
export CFLAGS="-I/opt/homebrew/include"
export LDFLAGS="-L/opt/homebrew/lib"
```

4. Try Installing with --no-cache-dir
Sometimes cached installations cause issues. You can try reinstalling with the --no-cache-dir flag:

`pip install spacy --no-cache-dir`

5. Consider Using Conda (Optional)
If you are using a conda environment (as your path seems to suggest), try using conda install to install spacy and its dependencies through Conda’s package manager, which may handle these dependencies more smoothly:

`conda install -c conda-forge spacy`

This would avoid the need to compile the package from source.

6. Update Cython
Ensure that your Cython installation is up-to-date, as outdated versions may cause issues with newer code:

`pip install --upgrade cython`

**Reply by eginhard** [reply]
@sapardo10 Try installing [our fork](https://github.com/idiap/coqui-ai-TTS) with `pip install coqui-tts`, this repo is not maintained anymore.

**Reply by sapardo10** [reply]
Thanks @eginhard, tried it and worked right out of the box 👍

**Reply by cosmocatalano** [reply]
Isn't the non-maintenance of this repo something that should be documented in a more prominent location?
